### ISSUE TITLE
SDL_RenderGeometryRaw: optionally copy additional coordinates based on stride

### ISSUE THREAD

**Issue by expikr** [libsdl-org/SDL#11276]
SDL_RenderGeometryRaw already lets you specify an arbitrary stride for your vertex list.

It would be very useful to allow up to four spatial coordinates to be transmitted to the render backend if the stride is greater than 8 bytes.

Backends which accept up to four spatial coordinates should have the third and fourth coordinate default to zero and one respectively if the user provides a stride of only two coordinates.

SDL should make no additional guarantee or responsibility beyond dumb transmission of the coordinates to the hardware, and the software renderer should do nothing at all with the third and fourth optional coordinate.

Example:
https://github.com/libsdl-org/SDL/blob/e0321ca5b6cb70ce178e533884b407571d2d1002/src/render/opengl/SDL_render_gl.c#L963
https://github.com/libsdl-org/SDL/blob/e0321ca5b6cb70ce178e533884b407571d2d1002/src/render/opengl/SDL_render_gl.c#L994-L995
```diff
--- size_t sz = 2 * sizeof(GLfloat) + 4 * sizeof(GLfloat) + (texture ? 2 : 0) * sizeof(GLfloat);
+++ size_t sz = 4 * sizeof(GLfloat) + 4 * sizeof(GLfloat) + (texture ? 2 : 0) * sizeof(GLfloat);
```

```diff
        *(verts++) = xy_[0] * scale_x;
        *(verts++) = xy_[1] * scale_y;
+++     *(verts++) = xy_stride >= 12 ? xy_[2] : 0;
+++     *(verts++) = xy_stride >= 16 ? xy_[3] : 1;
```

**Comment by slouken** [comment]
The stride could indicate some other data besides additional vertex coordinates. What's the use case for this?

**Comment by slouken** [comment]
Can you provide a little demo program?

**Comment by slime73** [comment]
> SDL should make no additional guarantee or responsibility beyond dumb transmission of the coordinates to the hardware

Wouldn't this break vertex positions for people who have wide strides that contain non-position data?

For example if they have an array of `{XY, RGBA, UV}` structs, and use the stride parameters to make sure the fields of the structs get sent to the right vertex inputs. That's a pretty common use case for strides currently I think.

**Comment by slouken** [comment]
> Wouldn't this break vertex positions for people who have wide strides that contain non-position data?

Yes, it would. I’m pretty sure the answer here is that you should use the new GPU API, I’m just curious what could be done with this.

**Comment by slouken** [comment]
That's interesting, but I mean from a practical point of view, for creating an SDL game.

**Comment by slouken** [comment]
I took a look at this a bit, and allowing this is not trivial. The assumption that you're only using 2 floats for xy position and that you're using the normal model view matrices is baked into all back ends.

You're probably better off writing a thin layer over the SDL GPU API that has this designed in, along with support for shaders and so forth. If you write something like that I'd love to see it!

**Comment by slouken** [comment]
Well, feel free to submit a proof of concept PR.

**Comment by slouken** [comment]
> Do you mean this as "you are welcome to try to add this functionality to the Render API by wrapping the GPU API", or did you meant it as "if you want to do something like this, you're better off going straight to the GPU API instead of Render API"?

I meant the latter. The implementation isn't that hard (you can look at SDL_render_gpu.c for inspiration) and you'll have complete flexibility over how the API works for you. You can add full coordinate support and shader support, etc.

**Comment by slouken** [comment]
> And you are still open for a PR demonstrating that the functionality is trivially added to the RenderAPI without altering any of the matrices currently used by the backends?

Sure.

**Comment by slouken** [comment]
> Is a fully functioning PR required to convince you that the existing matrices do not need to be touched in any way, or can a toy GL1.1 example matching the same assumptions made by the backend be enough to get the point across?

I understand the point, I just am not sure how the existing code could be easily modified to accept data in the way you propose. That's one of the reasons I'm suggesting you write a new API on top of the GPU API that gives you more flexibility and lets you pass through data however you choose. We often have people asking for the render API + stuff, and maybe it's time to write that. :)

> It is a lot of abstractions to dig through from scratch without existing knowledge of the organizational cruft just to prove a point, I'm concerned about whether there's a release timeline cutoff for preliminary consideration, which will require a hyper-accelerated effort on my part in order to meet.

No, there's no release timeline cutoff. If we added something like this it would not be using the proposed API, instead we would probably add a new function that allowed the application to specify the vertex layout, and that could be done anytime.

> In any case, it'd help if you can point me to where and which assumptions are made and where the geometries are actually communicated to the GPU.

Take a look at QueueCmdGeometry() in SDL_render.c, that's probably the best place to start.

**Comment by slouken** [comment]
Interesting. I wonder if this would cause performance issues for existing platforms. I'm guessing not? The SDL 2D render API is pretty lightweight relative to most 3D GPU loads.

**Comment by slouken** [comment]
Internals can change, and new API functions can be added, we just can't change the existing public API.
